cpp/src/io/utilities/datasource.hpp

```diff
@@ -87,6 +97,54 @@ class datasource {
    * @return bool True if there is data, False otherwise
    **/
   virtual bool empty() const { return size() == 0; }
+
+  /**
+   * @brief Returns whether the data source returns fixed-address buffers
+   *
+   * @return bool True if get_buffer addresses only depend on offset, False otherwise
+   **/
+  virtual bool has_fixed_mappings() const { return false; }
+
+  /**
+   * @brief Flush any pending read requests (for async sources)
+   *
+   **/
+  virtual void flush() {}
+
+  /**
+   * @brief Returns a device buffer with a subset of data from the source
+   *
+   * @param[in] offset Bytes from the start
+   * @param[in] size Bytes to read
+   * @param[in] stream cuda stream to use
+   *
+   * @return rmm::device_buffer the data buffer
+   **/
+  virtual rmm::device_buffer get_device_buffer(size_t offset, size_t size, cudaStream_t stream = 0) {
```

Comment:
Any reason to implement this in the header rather than the .cpp file like the other methods? Likewise `device_read`.

Reply:
Mainly because it's in the base non-constructible class (all others base class members were implemented in the header, so followed what seemed to be the local convention). I suppose I could move these to the cpp.